File: examples/entra-id/csharp/EntraIdAuthProvider.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using Cassandra;
using Azure.Identity;
using Microsoft.Identity.Client;

namespace Cassandra.Auth.EntraId
{
    /// <summary>
    /// DataStax C# Driver IAuthProvider implementation for Microsoft Entra ID token-based authentication.
    /// 
    /// This provider acquires a JWT access token from Microsoft Entra ID using MSAL.NET
    /// (Microsoft.Identity.Client) or Azure.Identity and sends it to Apache Cassandra's
    /// EntraIdAuthenticator via the SASL PLAIN mechanism.
    /// 
    /// <example>
    /// Usage with Client Credentials (Service Principal):
    /// <code>
    /// var authProvider = new EntraIdAuthProvider(
    ///     tenantId: "your-tenant-id",
    ///     clientId: "your-client-id",
    ///     clientSecret: "your-client-secret"
    /// );
    /// 
    /// var cluster = Cluster.Builder()
    ///     .AddContactPoint("cassandra-host")
    ///     .WithAuthProvider(authProvider)
    ///     .Build();
    /// </code>
    /// </example>
    /// 
    /// <example>
    /// Usage with DefaultAzureCredential (recommended for Azure-hosted apps):
    /// <code>
    /// var authProvider = EntraIdAuthProvider.WithDefaultCredential(
    ///     tenantId: "your-tenant-id",
    ///     clientId: "your-client-id"
    /// );
    /// </code>
    /// </example>
    /// 
    /// NuGet Dependencies:
    /// - CassandraCSharpDriver (>= 3.19.0)
    /// - Microsoft.Identity.Client (>= 4.56.0)
    /// - Azure.Identity (>= 1.10.0)  [optional, for DefaultAzureCredential]
    /// </summary>
    public class EntraIdAuthProvider : IAuthProvider
    {
        private readonly string _tenantId;
        private readonly string _clientId;
        private readonly string _clientSecret;
        private readonly string _username;
        private readonly string _password;
        private readonly string _scope;
        private readonly string _authority;
        private readonly bool _useManagedIdentity;
        private readonly bool _useDefaultCredential;

        private IConfidentialClientApplication _confidentialClient;
        private IPublicClientApplication _publicClient;
        private readonly object _lock = new object();

        /// <summary>
        /// Creates an EntraIdAuthProvider with client credentials (service principal) flow.
        /// </summary>
        /// <param name="tenantId">Microsoft Entra ID tenant ID (directory ID)</param>
        /// <param name="clientId">Application (client) ID registered in Entra ID</param>
        /// <param name="clientSecret">Client secret for the application</param>
        /// <param name="scope">OAuth2 scope. Default: api://{clientId}/.default</param>
        public EntraIdAuthProvider(string tenantId, string clientId, string clientSecret, string scope = null)
        {
            _tenantId = tenantId ?? throw new ArgumentNullException(nameof(tenantId));
            _clientId = clientId ?? throw new ArgumentNullException(nameof(clientId));
            _clientSecret = clientSecret ?? throw new ArgumentNullException(nameof(clientSecret));
            _scope = scope ?? $"api://{clientId}/.default";
            _authority = $"https://login.microsoftonline.com/{tenantId}";
        }

        /// <summary>
        /// Creates an EntraIdAuthProvider with username/password (ROPC) flow.
        /// WARNING: ROPC is not recommended for production. Use client credentials or managed identity.
        /// </summary>
        public EntraIdAuthProvider(string tenantId, string clientId, string username, string password,
                                   string scope = null, bool isRopc = true)
        {
            _tenantId = tenantId ?? throw new ArgumentNullException(nameof(tenantId));
            _clientId = clientId ?? throw new ArgumentNullException(nameof(clientId));
            _username = username ?? throw new ArgumentNullException(nameof(username));
            _password = password ?? throw new ArgumentNullException(nameof(password));
            _scope = scope ?? $"api://{clientId}/.default";
            _authority = $"https://login.microsoftonline.com/{tenantId}";
        }

        /// <summary>
        /// Private constructor for managed identity / default credential flows.
        /// </summary>
        private EntraIdAuthProvider(string tenantId, string clientId, bool useManagedIdentity,
                                    bool useDefaultCredential, string scope)
        {
            _tenantId = tenantId;
            _clientId = clientId;
            _scope = scope ?? $"api://{clientId}/.default";
            _authority = $"https://login.microsoftonline.com/{tenantId}";
            _useManagedIdentity = useManagedIdentity;
            _useDefaultCredential = useDefaultCredential;
        }

        /// <summary>
        /// Creates an EntraIdAuthProvider using Azure Managed Identity.
        /// Use when running on Azure VMs, App Service, AKS, etc.
        /// </summary>
        public static EntraIdAuthProvider WithManagedIdentity(string tenantId, string clientId, string scope = null)
        {
            return new EntraIdAuthProvider(tenantId, clientId,
                useManagedIdentity: true, useDefaultCredential: false, scope: scope);
        }

        /// <summary>
        /// Creates an EntraIdAuthProvider using DefaultAzureCredential.
        /// This automatically tries multiple credential types (environment, managed identity,
        /// VS Code, Azure CLI, etc.) - recommended for Azure-hosted applications.
        /// </summary>
        public static EntraIdAuthProvider WithDefaultCredential(string tenantId, string clientId, string scope = null)
        {
            return new EntraIdAuthProvider(tenantId, clientId,
                useManagedIdentity: false, useDefaultCredential: true, scope: scope);
        }

        /// <summary>
        /// Creates a new authenticator for each connection.
        /// </summary>
        public IAuthenticator NewAuthenticator(IPEndPoint host)
        {
            return new EntraIdAuthenticator(this);
        }

        /// <summary>
        /// Acquires a JWT access token from Microsoft Entra ID.
        /// </summary>
        internal string AcquireToken()
        {
            if (_useDefaultCredential)
            {
                return AcquireTokenWithDefaultCredential();
            }
            else if (_useManagedIdentity)
            {
                return AcquireTokenWithManagedIdentity();
            }
            else if (!string.IsNullOrEmpty(_clientSecret))
            {
                return AcquireTokenWithClientCredentials();
            }
            else if (!string.IsNullOrEmpty(_username))
            {
                return AcquireTokenWithUsernamePassword();
            }
            else
            {
                throw new AuthenticationException("No authentication method configured.");
            }
        }

        private string AcquireTokenWithClientCredentials()
        {
            if (_confidentialClient == null)
            {
                lock (_lock)
                {
                    if (_confidentialClient == null)
                    {
                        _confidentialClient = ConfidentialClientApplicationBuilder.Create(_clientId)
                            .WithClientSecret(_clientSecret)
                            .WithAuthority(new Uri(_authority))
                            .Build();
                    }
                }
            }

            var result = _confidentialClient.AcquireTokenForClient(new[] { _scope })
                .ExecuteAsync().GetAwaiter().GetResult();
            return result.AccessToken;
        }

        private string AcquireTokenWithUsernamePassword()
        {
            if (_publicClient == null)
            {
                lock (_lock)
                {
                    if (_publicClient == null)
                    {
                        _publicClient = PublicClientApplicationBuilder.Create(_clientId)
                            .WithAuthority(new Uri(_authority))
                            .Build();
                    }
                }
            }

            var result = _publicClient.AcquireTokenByUsernamePassword(new[] { _scope }, _username, _password)
                .ExecuteAsync().GetAwaiter().GetResult();
            return result.AccessToken;
        }

        private string AcquireTokenWithManagedIdentity()
        {
            var miApp = ManagedIdentityApplicationBuilder
                .Create(ManagedIdentityId.SystemAssigned)
                .Build();

            var result = miApp.AcquireTokenForManagedIdentity(_scope)
                .ExecuteAsync().GetAwaiter().GetResult();
            return result.AccessToken;
        }

        private string AcquireTokenWithDefaultCredential()
        {
            // Uses Azure.Identity DefaultAzureCredential
            var credential = new DefaultAzureCredential(new DefaultAzureCredentialOptions
            {
                TenantId = _tenantId
            });

            var tokenRequestContext = new Azure.Core.TokenRequestContext(new[] { _scope });
            var token = credential.GetToken(tokenRequestContext);
            return token.Token;
        }

        /// <summary>
        /// Internal authenticator that handles SASL PLAIN negotiation with Entra ID JWT tokens.
        /// </summary>
        private class EntraIdAuthenticator : IAuthenticator
        {
            private readonly EntraIdAuthProvider _provider;

            public EntraIdAuthenticator(EntraIdAuthProvider provider)
            {
                _provider = provider;
            }

            /// <summary>
            /// Returns the initial SASL PLAIN response containing the JWT token.
            /// Format: authzId NUL authnId NUL password(JWT)
            /// </summary>
            public byte[] InitialResponse()
            {
                string token = _provider.AcquireToken();
                string authnId = !string.IsNullOrEmpty(_provider._username)
                    ? _provider._username
                    : "entra-token";

                byte[] authnIdBytes = Encoding.UTF8.GetBytes(authnId);
                byte[] tokenBytes = Encoding.UTF8.GetBytes(token);

                byte[] response = new byte[1 + authnIdBytes.Length + 1 + tokenBytes.Length];
                // authzId (empty) NUL
                response[0] = 0;
                // authnId
                Buffer.BlockCopy(authnIdBytes, 0, response, 1, authnIdBytes.Length);
                // NUL separator
                response[1 + authnIdBytes.Length] = 0;
                // password (JWT token)
                Buffer.BlockCopy(tokenBytes, 0, response, 2 + authnIdBytes.Length, tokenBytes.Length);

                return response;
            }

            public byte[] EvaluateChallenge(byte[] challenge)
            {
                // SASL PLAIN has no challenge-response
                return null;
            }
        }
    }
}

File: examples/entra-id/java/EntraIdAuthProvider.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.example.cassandra.auth;

import com.datastax.driver.core.AuthProvider;
import com.datastax.driver.core.Authenticator;
import com.datastax.driver.core.exceptions.AuthenticationException;
import com.microsoft.aad.msal4j.*;

import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.concurrent.CompletableFuture;

/**
 * DataStax Java Driver 3.x AuthProvider for Microsoft Entra ID token-based authentication.
 *
 * <p>This provider acquires a JWT access token from Microsoft Entra ID using MSAL4J
 * and sends it to Apache Cassandra's {@code EntraIdAuthenticator} via the SASL PLAIN mechanism.</p>
 *
 * <h3>Usage with Client Credentials (Service Principal / Managed Identity):</h3>
 * <pre>{@code
 * EntraIdAuthProvider authProvider = EntraIdAuthProvider.builder()
 *     .tenantId("your-tenant-id")
 *     .clientId("your-client-id")
 *     .clientSecret("your-client-secret")
 *     .scope("api://your-client-id/.default")
 *     .build();
 *
 * Cluster cluster = Cluster.builder()
 *     .addContactPoint("cassandra-host")
 *     .withAuthProvider(authProvider)
 *     .build();
 * }</pre>
 *
 * <h3>Usage with Username/Password (ROPC flow - for dev/test only):</h3>
 * <pre>{@code
 * EntraIdAuthProvider authProvider = EntraIdAuthProvider.builder()
 *     .tenantId("your-tenant-id")
 *     .clientId("your-client-id")
 *     .username("user@contoso.com")
 *     .password("user-password")
 *     .scope("api://your-client-id/.default")
 *     .build();
 * }</pre>
 *
 * <h3>Maven Dependencies:</h3>
 * <pre>{@code
 * <dependency>
 *     <groupId>com.datastax.cassandra</groupId>
 *     <artifactId>cassandra-driver-core</artifactId>
 *     <version>3.11.5</version>
 * </dependency>
 * <dependency>
 *     <groupId>com.microsoft.azure</groupId>
 *     <artifactId>msal4j</artifactId>
 *     <version>1.14.0</version>
 * </dependency>
 * }</pre>
 *
 * @see <a href="https://learn.microsoft.com/en-us/entra/identity-platform/">Microsoft Entra ID Platform</a>
 */
public class EntraIdAuthProvider implements AuthProvider
{
    private final String tenantId;
    private final String clientId;
    private final String clientSecret;      // For client_credentials flow
    private final String username;           // For ROPC flow
    private final String password;           // For ROPC flow
    private final String scope;
    private final String authority;
    private final boolean useManagedIdentity;

    private volatile IConfidentialClientApplication confidentialClient;
    private volatile IPublicClientApplication publicClient;

    private EntraIdAuthProvider(Builder builder)
    {
        this.tenantId = builder.tenantId;
        this.clientId = builder.clientId;
        this.clientSecret = builder.clientSecret;
        this.username = builder.username;
        this.password = builder.password;
        this.scope = builder.scope != null ? builder.scope : "api://" + clientId + "/.default";
        this.authority = builder.authority != null
            ? builder.authority
            : "https://login.microsoftonline.com/" + tenantId;
        this.useManagedIdentity = builder.useManagedIdentity;
    }

    @Override
    public Authenticator newAuthenticator(InetSocketAddress host, String authenticator)
        throws AuthenticationException
    {
        // The authenticator string is the FQCN of the server-side authenticator
        // We only respond to EntraIdAuthenticator or any authenticator (for flexibility)
        return new EntraIdAuthenticator();
    }

    /**
     * Acquires an access token from Microsoft Entra ID.
     * Supports three flows:
     * 1. Client Credentials (service principal) - when clientSecret is set
     * 2. Resource Owner Password Credentials (ROPC) - when username/password are set
     * 3. Managed Identity - when useManagedIdentity is true
     */
    private String acquireToken()
    {
        try
        {
            if (useManagedIdentity)
            {
                return acquireTokenWithManagedIdentity();
            }
            else if (clientSecret != null && !clientSecret.isEmpty())
            {
                return acquireTokenWithClientCredentials();
            }
            else if (username != null && !username.isEmpty())
            {
                return acquireTokenWithUsernamePassword();
            }
            else
            {
                throw new AuthenticationException(
                    null, "No authentication method configured. " +
                    "Provide clientSecret (service principal), username/password (ROPC), " +
                    "or enable Managed Identity.");
            }
        }
        catch (AuthenticationException e)
        {
            throw e;
        }
        catch (Exception e)
        {
            throw new AuthenticationException(null,
                "Failed to acquire Entra ID token: " + e.getMessage());
        }
    }

    private String acquireTokenWithClientCredentials() throws Exception
    {
        if (confidentialClient == null)
        {
            synchronized (this)
            {
                if (confidentialClient == null)
                {
                    confidentialClient = ConfidentialClientApplication.builder(clientId,
                            ClientCredentialFactory.createFromSecret(clientSecret))
                        .authority(authority)
                        .build();
                }
            }
        }

        ClientCredentialParameters parameters = ClientCredentialParameters.builder(
            Collections.singleton(scope))
            .build();

        IAuthenticationResult result = confidentialClient.acquireToken(parameters).get();
        return result.accessToken();
    }

    private String acquireTokenWithUsernamePassword() throws Exception
    {
        if (publicClient == null)
        {
            synchronized (this)
            {
                if (publicClient == null)
                {
                    publicClient = PublicClientApplication.builder(clientId)
                        .authority(authority)
                        .build();
                }
            }
        }

        UserNamePasswordParameters parameters = UserNamePasswordParameters.builder(
            Collections.singleton(scope), username, password.toCharArray())
            .build();

        IAuthenticationResult result = publicClient.acquireToken(parameters).get();
        return result.accessToken();
    }

    private String acquireTokenWithManagedIdentity() throws Exception
    {
        // For Managed Identity, use the ManagedIdentityApplication
        ManagedIdentityApplication miApp = ManagedIdentityApplication.builder(
            ManagedIdentityId.systemAssigned())
            .build();

        ManagedIdentityParameters parameters = ManagedIdentityParameters.builder(scope)
            .build();

        IAuthenticationResult result = miApp.acquireTokenForManagedIdentity(parameters).get();
        return result.accessToken();
    }

    /**
     * SASL PLAIN authenticator that sends the Entra ID JWT token as the password.
     * The SASL PLAIN format is: authzId NUL authnId NUL password
     * where password contains the JWT access token.
     */
    private class EntraIdAuthenticator implements Authenticator
    {
        private static final byte NUL = 0;

        @Override
        public byte[] initialResponse()
        {
            String token = acquireToken();
            // Determine the username to send (informational - server derives identity from JWT)
            String authnId = (username != null && !username.isEmpty()) ? username : "entra-token";

            // SASL PLAIN format: authzId<NUL>authnId<NUL>password
            byte[] authnIdBytes = authnId.getBytes(StandardCharsets.UTF_8);
            byte[] tokenBytes = token.getBytes(StandardCharsets.UTF_8);

            ByteBuffer buffer = ByteBuffer.allocate(1 + authnIdBytes.length + 1 + tokenBytes.length);
            buffer.put(NUL);                 // authzId (empty)
            buffer.put(authnIdBytes);        // authnId (username)
            buffer.put(NUL);                 // separator
            buffer.put(tokenBytes);          // password (JWT token)

            return buffer.array();
        }

        @Override
        public byte[] evaluateChallenge(byte[] challenge)
        {
            // SASL PLAIN has no challenge-response; this should not be called
            return null;
        }

        @Override
        public void onAuthenticationSuccess(byte[] token)
        {
            // Authentication completed successfully
        }
    }

    // -------------------------
    // Builder
    // -------------------------

    public static Builder builder()
    {
        return new Builder();
    }

    public static class Builder
    {
        private String tenantId;
        private String clientId;
        private String clientSecret;
        private String username;
        private String password;
        private String scope;
        private String authority;
        private boolean useManagedIdentity = false;

        public Builder tenantId(String tenantId)
        {
            this.tenantId = tenantId;
            return this;
        }

        public Builder clientId(String clientId)
        {
            this.clientId = clientId;
            return this;
        }

        public Builder clientSecret(String clientSecret)
        {
            this.clientSecret = clientSecret;
            return this;
        }

        public Builder username(String username)
        {
            this.username = username;
            return this;
        }

        public Builder password(String password)
        {
            this.password = password;
            return this;
        }

        /**
         * Sets the scope for token acquisition.
         * Default: api://{clientId}/.default
         */
        public Builder scope(String scope)
        {
            this.scope = scope;
            return this;
        }

        /**
         * Sets the Microsoft Entra ID authority URL.
         * Default: https://login.microsoftonline.com/{tenantId}
         */
        public Builder authority(String authority)
        {
            this.authority = authority;
            return this;
        }

        /**
         * Enables Azure Managed Identity authentication.
         * Use when running on Azure VMs, App Service, AKS, etc.
         */
        public Builder useManagedIdentity(boolean useManagedIdentity)
        {
            this.useManagedIdentity = useManagedIdentity;
            return this;
        }

        public EntraIdAuthProvider build()
        {
            if (tenantId == null || tenantId.isEmpty())
                throw new IllegalArgumentException("tenantId is required");
            if (clientId == null || clientId.isEmpty())
                throw new IllegalArgumentException("clientId is required");
            return new EntraIdAuthProvider(this);
        }
    }
}
